src/nodes.coffee

```diff
@@ -3551,6 +3553,33 @@ exports.Code = class Code extends Base
 
     seenSuper
 
+  propagateLhs: ->
+    for {name} in @params when name instanceof Arr or name instanceof Obj
+      name.propagateLhs yes
```

Comment:
Make destructured array/object params aware that they're left-hand-side (so that they'll eg generate AST nodes of type `ObjectPattern` instead of `ObjectExpression`) with the existing pattern using `propagateLhs()`